File: Day2/2-2.hs
data Direction = U | R | D | L deriving (Show, Read)
type Command = [Direction]
type Instructions = [Command]

kA :: Int
kA = 10
kB :: Int
kB = 11
kC :: Int
kC = 12
kD :: Int
kD = 13

getNumber :: Int -> Command -> Int
getNumber n [] = n
getNumber n (U:cs) = getNumber nextNum cs
                     where nextNum = goUp n
getNumber n (R:cs) = getNumber nextNum cs
                     where nextNum = goRight n
getNumber n (D:cs) = getNumber nextNum cs
                     where nextNum = goDown n
getNumber n (L:cs) = getNumber nextNum cs
                     where nextNum = goLeft n

goUp :: Int -> Int
goUp  3 = 1
goUp  6 = 2
goUp  7 = 3
goUp  8 = 4
goUp n
  | n == kA = 6
  | n == kB = 7
  | n == kC = 8
  | n == kD = kB
  | otherwise = n

goDown :: Int -> Int
goDown 1 =  3
goDown 2 =  6
goDown 3 =  7
goDown 4 =  8
goDown 6 = kA
goDown 7 = kB
goDown 8 = kC
goDown n
  | n == kB = kD
  | otherwise = n

goLeft :: Int -> Int
goLeft 3 = 2
goLeft 4 = 3
goLeft n
  | n >= 6 && n <= 9 = n - 1
  | n == kB = kA
  | n == kC = kB
  | otherwise = n

goRight :: Int -> Int
goRight 2 = 3
goRight 3 = 4
goRight n
  | n >= 5 && n <= 8 = n + 1
  | n == kA = kB
  | n == kB = kC
  | otherwise = n

readInstructions :: Int -> Instructions -> [Int]
readInstructions _ [] = []
readInstructions n (i:is) = let num = getNumber n i
                            in num : readInstructions num is

main :: IO ()
main = do
  strInstr <- fmap lines $ readFile "input.txt"
  let instr = map (\a -> map (\b -> read [b] :: Direction) a) strInstr
  putStrLn $ show $ readInstructions 5 instr

File: Day02_JavaScript/2-2.js
var fs = require('fs');

function moveUp(n) {
    if (n === '3') return '1';
    else if (n === '6') return '2';
    else if (n === '7') return '3';
    else if (n === '8') return '4';
    else if (n === 'A') return '6';
    else if (n === 'B') return '7';
    else if (n === 'C') return '8';
    else if (n === 'D') return 'B';
    else return n;
}

function moveDown(n) {
    if (n === '1') return '3';
    else if (n === '2') return '6';
    else if (n === '3') return '7';
    else if (n === '4') return '8';
    else if (n === '6') return 'A';
    else if (n === '7') return 'B';
    else if (n === '8') return 'C';
    else if (n === 'B') return 'D';
    else return n;
}

function moveRight(n) {
    if (n === '2') return '3';
    else if (n === '3') return '4';
    else if (n === '5') return '6';
    else if (n === '6') return '7';
    else if (n === '7') return '8';
    else if (n === '8') return '9';
    else if (n === 'A') return 'B';
    else if (n === 'B') return 'C';
    else return n;
}

function moveLeft(n) {
    if (n === '3') return '2';
    else if (n === '4') return '3';
    else if (n === '6') return '5';
    else if (n === '7') return '6';
    else if (n === '8') return '7';
    else if (n === '9') return '8';
    else if (n === 'B') return 'A';
    else if (n === 'C') return 'B';
    else return n;
}

function getNumber(n, commands) {
    var num = n;
    for (var i = 0, len = commands.length; i < len; i++) {
        var command = commands.charAt(i);
        if (command == 'U') num = moveUp(num);
        else if (command == 'D') num = moveDown(num);
        else if (command == 'R') num = moveRight(num);
        else if (command == 'L') num = moveLeft(num);
    }
    return num;
}

function readInstructions(instructions) {
    var code = "";
    var num = '5';
    for (var i = 0, len = instructions.length; i < len; i++) {
        num = getNumber(num, instructions[i]);
        code += num;
    }
    return code;
}

function solveCodeFromInput(input) {
    var lines = input.match(/[^\r\n]+/g);
    return readInstructions(lines);
}

fs.readFile('input.txt', 'utf8', function(err, input) {
    if (err)
        console.log('Could not read input file.');
    else
        console.log(solveCodeFromInput(input));
});
